core/src/main/java/com/datastax/oss/driver/api/core/config/DriverConfigLoader.java

```diff
@@ -16,14 +16,123 @@
 package com.datastax.oss.driver.api.core.config;
 
 import com.datastax.oss.driver.api.core.context.DriverContext;
+import com.datastax.oss.driver.api.core.session.SessionBuilder;
+import com.datastax.oss.driver.internal.core.config.typesafe.DefaultDriverConfigLoader;
+import com.typesafe.config.Config;
+import com.typesafe.config.ConfigFactory;
 import edu.umd.cs.findbugs.annotations.NonNull;
+import java.io.File;
+import java.net.URL;
 import java.util.concurrent.CompletionStage;
 
 /**
  * Manages the initialization, and optionally the periodic reloading, of the driver configuration.
+ *
+ * @see SessionBuilder#withConfigLoader(DriverConfigLoader)
  */
 public interface DriverConfigLoader extends AutoCloseable {
 
+  /**
+   * Builds an instance using the driver's default implementation (based on Typesafe config), except
+   * that application-specific options are loaded from a classpath resource with a custom name.
+   *
+   * <p>More precisely, configuration properties are loaded and merged from the following
+   * (first-listed are higher priority):
+   *
+   * <ul>
+   *   <li>system properties
+   *   <li>{@code <resourceBaseName>.conf} (all resources on classpath with this name)
+   *   <li>{@code <resourceBaseName>.json} (all resources on classpath with this name)
+   *   <li>{@code <resourceBaseName>.properties} (all resources on classpath with this name)
+   *   <li>{@code reference.conf} (all resources on classpath with this name). In particular, this
```

Comment:
What will happen if the user will create a file `reference.conf` and put it on it's classpath?

Reply:
The two files are merged, the order of precedence is undefined.

There is no reason for a client application to do that. One valid case where it might happen is if another dependency is using Typesafe config, but in that case the options won't collide thanks to the `datastax-java-driver` prefix.